jsonrpc/dispatcher_test.go

```diff
@@ -266,25 +266,86 @@ func TestDispatcherFuncDecode(t *testing.T) {
-    {"id":2,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x2", true]},
-    {"id":3,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x3", true]},
-	{"id":4,"jsonrpc":"2.0","method": "web3_sha3","params": ["0x68656c6c6f20776f726c64"]}
-]`)...))
-	assert.NoError(t, err)
-
-	var res []SuccessResponse
-
-	assert.NoError(t, expectBatchJSONResult(resp, &res))
-	assert.Len(t, res, 4)
+		return res
+	}
 
-	jsonerr := &ObjectError{Code: -32602, Message: "Invalid Params"}
+	cases := []struct {
+		id          uint8
+		desc        string
+		dispatcher  *Dispatcher
+		reqBody     []byte
+		expectError bool
+	}{
+		{
+			1,
+			"test with leading whitespace (\"  \\t\\n\\n\\r\\)",
+			newDispatcher(hclog.NewNullLogger(), newMockStore(), 0, 0, 20, 1000),
+
+			append([]byte{0x20, 0x20, 0x09, 0x0A, 0x0A, 0x0D}, []byte(`[
+				{"id":1,"jsonrpc":"2.0","method":"eth_getBalance","params":["0x1", true]},
+				{"id":2,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x2", true]},
+				{"id":3,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x3", true]},
+				{"id":4,"jsonrpc":"2.0","method": "web3_sha3","params": ["0x68656c6c6f20776f726c64"]}]`)...),
+			false,
+		},
+		{
+			2,
+			"test with batch req length within batchRequestLengthLimit",
+			newDispatcher(hclog.NewNullLogger(), newMockStore(), 0, 0, 10, 1000),
+			[]byte(`[
+				{"id":1,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":2,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":3,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":4,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":5,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":6,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]}]`),
+			false,
+		},
+		{
+			3,
+			"test with batch req length exceeding batchRequestLengthLimit",
+			newDispatcher(hclog.NewNullLogger(), newMockStore(), 0, 0, 3, 1000),
+			[]byte(`[
+				{"id":1,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":2,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":3,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":4,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":5,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]},
+				{"id":6,"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["latest", true]}]`),
+			true,
+		},
+	}
 
-	assert.Equal(t, res[0].Error, jsonerr)
-	assert.Nil(t, res[3].Error)
+	for _, c := range cases {
+		res := handle(c.dispatcher, c.reqBody)
+
+		if c.expectError {
+			var resp ErrorResponse
+
+			assert.NoError(t, expectBatchJSONResult(res, &resp))
+
+			switch c.id {
+			case 3:
```

Comment:
I think using switch-case by test id is not best practice. All information that needs test can be given in `cases` data